trie/iter.py

```diff
@@ -1,89 +1,127 @@
-from trie.constants import (
-    NODE_TYPE_BLANK,
-    NODE_TYPE_LEAF,
-    NODE_TYPE_EXTENSION,
-    NODE_TYPE_BRANCH,
-)
 from trie.utils.nibbles import (
     bytes_to_nibbles,
     nibbles_to_bytes,
-    remove_nibbles_terminator,
 )
 from trie.utils.nodes import (
-    get_node_type,
-    extract_key,
-    key_starts_with,
+    consume_common_prefix,
+)
+from trie.exceptions import (
+    PerfectVisibility,
+)
+from trie.fog import (
+    HexaryTrieFog,
+    TrieFrontierCache,
 )
 
 
-# FIXME: This is probably very inneficient and the code itself is quite convoluted. We should
-# either look into refactoring or rewriting it from scratch eventually.
 class NodeIterator:
     """Iterate over all nodes of a trie, ensuring its consistency."""
 
     def __init__(self, trie):
-        self.trie = trie
+        self._trie = trie
 
-    def next(self, key):
-        key = bytes_to_nibbles(key)
-        nibbles = self._iter(self.trie.root_node, key)
-        if nibbles is None:
-            return None
-        return nibbles_to_bytes(remove_nibbles_terminator(nibbles))
+    def next(self, key_bytes):
+        """
+        Find the next key to the right from the given key, or None if there is
+        no key to the right.
+
+        .. NOTE:: If you plan to iterate the full trie, use all() instead, for performance.
 
-    def _get_next(self, node):
-        node_type = get_node_type(node)
-        if node_type == NODE_TYPE_BLANK:
+        :return: key in bytes to the right of key_bytes, or None
+        """
+        key = bytes_to_nibbles(key_bytes)
+        traversed = ()
+        root = self._trie.traverse(traversed)
```

Comment:
This feels a bit convoluted if all we want is to get the root. Isn't there a simper way?

Reply:
Yes! Now that #106 is merged...

I will update it.